### ISSUE TITLE
In certain cases it fails to find the pass entry

### ISSUE THREAD

**Issue by divansantana** [DamienCassou/auth-password-store#17]
Similar to #11 , there are times when `auth-password-store` fails to find an entry.

From what I can see the pass auth is setup correctly, so not exactly sure why it doesn't find it.

Example:

I use `eshell` with and do this `cd "/ssh:htf-sentry|sudo:root@htf-sentry:/root"`
```
Tramp: Opening connection for root@htf-sentry using sudo...
Tramp: Sending command ‘exec ssh    -e none htf-sentry’
Tramp: Waiting for prompts from remote shell...done
Tramp: Found remote shell prompt on ‘htf-sentry’
Tramp: Sending command ‘exec env SHELL=/bin/sh sudo -u root -s -H -p P""a""s""s""w""o""r""d"":’
Tramp: Waiting for prompts from remote shell...
Tramp: Sending Password
auth-password-store: searching for ’htf-sentry’ in entry names including user
auth-password-store: searching for ’htf-sentry’ in entry names not including user
auth-password-store: no entry name matched ’htf-sentry’, looking inside entries
auth-source-search: found 0 results (max 1) matching (:max 1 :user "root" :host "htf-sentry" :port "sudo")
Tramp: Waiting for prompts from remote shell...done
```
However I do have an entry in my pass name `htf-sentry`.
```
~  ᐅ l .password-store/htf-sentry.gpg 
-rw------- 1 admin admin 1.2K Sep 19 16:11 .password-store/htf-sentry.gpg
~  ᐅ pass show htf-sentry|head -n 1
ThePassword
~  ᐅ 
```

**Comment by divansantana** [comment]
> I have the same problem with dired-toggle-sudo.

Can't help much myself from a coding perspective. Maybe one day. :smile: 

If you need me to do any testing etc, just shout.

**Comment by DamienCassou** [comment]
@divansantana what you say regarding IRC seems to be a different issue. If you agree, do you mind opening a new issue?

If you can't help me with fixing, you may help me with a test. Have a look at https://github.com/DamienCassou/auth-password-store/blob/master/test/auth-password-store-tests.el#L64. If you can implement a test for the behavior you expect for the IRC case, that would be a great first contribution.

**Comment by divansantana** [comment]
> @divansantana what you say regarding IRC seems to be a different issue. If you agree, do you mind opening a new issue?

Absolutely. Yeah it should have been a different issue. I see you opened one there.

> If you can't help me with fixing, you may help me with a test. Have a look at https://github.com/DamienCassou/auth-password-store/blob/master/test/auth-password-store-tests.el#L64. If you can implement a test for the behavior you expect for the IRC case, that would be a great first contribution.

I'd like to say I can. But I have very little dev skills in any language. Esp elisp. And even less time. Though hope to be able to one day.

**Comment by divansantana** [comment]
Update on this and better description of the issue I'm seeing. Talking only about the original ssh/tramp and not ERC.

Tested using your latest commits in `match-subdomains` branch.

Using config `.init.el`
```emacs-lisp
(require 'package)
(setq package-enable-at-startup nil)
(setq package-archives '(			 
                         ("gnu"          . "https://elpa.gnu.org/packages/")
                         ("melpa"        . "https://melpa.org/packages/")
                         ))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

  (use-package password-store
    :ensure t
    :pin melpa)
  (use-package auth-password-store
    :ensure t
    :load-path "/home/admin/git/github/auth-password-store"
    :config
    (setq auth-source-debug t)
  (auth-pass-enable))
```

And pass entry
```
$ pass show htf-sentry
mypassword
```
Then open eshell and typing this:
```
~ $ cd "/ssh:htf-sentry|sudo:root@htf-sentry:/root"
```

The above fails to find the password for the sudo to root part. It prompts me as per below:
```
Tramp: Opening connection for root@htf-sentry using sudo...
Tramp: Sending command ‘exec ssh    -e none htf-sentry’
Tramp: Waiting for prompts from remote shell...done
Tramp: Found remote shell prompt on ‘htf-sentry’
Tramp: Sending command ‘exec env SHELL=/bin/sh sudo -u root -s -H -p P""a""s""s""w""o""r""d"":’
Tramp: Waiting for prompts from remote shell...
Tramp: Sending Password
auth-password-store: searching for ’htf-sentry’ in entry names including user
auth-source-search: found 0 results (max 1) matching (:max 1 :user "root" :host "htf-sentry" :port "sudo")
```

If I update my pass entry to include the `user` bit it then works, like so:
```
$ pass show htf-sentry
mypassword
user:root
```

output
```
Tramp: Opening connection for root@htf-sentry using sudo...
Tramp: Sending command ‘exec ssh    -e none htf-sentry’
Tramp: Waiting for prompts from remote shell...done
Tramp: Found remote shell prompt on ‘htf-sentry’
Tramp: Sending command ‘exec env SHELL=/bin/sh sudo -u root -s -H -p P""a""s""s""w""o""r""d"":’
Tramp: Waiting for prompts from remote shell...
Tramp: Sending Password
auth-password-store: searching for ’htf-sentry’ in entry names including user
auth-source-search: found 1 results (max 1) matching (:max 1 :user "root" :host "htf-sentry" :port "sudo")
Tramp: Waiting for prompts from remote shell...done
Tramp: Found remote shell prompt on ‘htf-sentry’
Tramp: Opening connection for root@htf-sentry using sudo...done
```

Specifying the `sudo` bit in the pass entry doesn't seem to change much.

Ideally I would like it to work with the pass entry containing only the password and not having to specify the user.

The reason or use case for me is that I have a common work password that is used for various services. I don't want to have to have multiple pass entries containing the same password. So I symlink the multiple entries to one common pass entry. The usernames/ports/urls etc may differ, but the actual password will be the same. So I'd prefer it to work with just the password entry. tramp/eshell emacs already know's the username.

Hope the above is clear.

**Comment by DamienCassou** [comment]
Thank you very much for your deep analyses. This will help me.

**Comment by DamienCassou** [comment]
Can you please test #19? You can use the `next` branch.

**Comment by divansantana** [comment]
Thanks for this.

I've tested the `massive-refactor` branch.

Overall it seems to be working extremely well and a lot faster/better.

The tramp/ssh+eshell issue works perfectly now. Also tested my usual
authentication sources, mediawiki, erc etc, all working very well. :smile:

Having an issue with mu4e, sending email which uses `smtpmail-send-it`. Details
below:

Trying to send an email with this configuration with master/melpa.

```emacs-lisp
(require 'package)
(setq package-enable-at-startup nil)
(setq package-archives '(
                         ("gnu"          . "https://elpa.gnu.org/packages/")
                         ("melpa"        . "https://melpa.org/packages/")
                         ))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

  (use-package password-store
    :ensure t
    :pin melpa)
  (use-package auth-password-store
    :ensure t
    :load-path "/home/admin/git/github/auth-password-store"
    :config
    (setq auth-source-debug t)
    (auth-pass-enable))

;; testing mail
  (use-package mu4e
    :load-path "/usr/share/emacs/site-lisp/mu4e"
    :config
    (setq mu4e-maildir "~/mail"
          user-full-name "Divan Santana"
          mu4e-user-mail-address-list '("dsantana@domain.co.za" "divan@home_domain.co.za")
          mu4e-get-mail-command "offlineimap -o -a Personal,Work -f INBOX"
          mu4e-update-interval 240
          mu4e-headers-auto-update t
          mu4e-compose-complete-only-after "2015-01-01"
          mu4e-compose-complete-only-personal t
          mu4e-headers-date-format "%Y-%m-%d"
          mu4e-attachment-dir  "/tmp"
          mu4e-view-show-images t
          message-send-mail-function 'smtpmail-send-it
          message-kill-buffer-on-exit t
          message-kill-buffer-query nil ;; don't prompt for kill draft
          mu4e-headers-leave-behavior 'apply ;; don't prompt just do
          mu4e-confirm-quit nil;; don't prompt for mu4e quit
          mu4e-compose-signature-auto-include t
          mu4e-compose-dont-reply-to-self t
          mu4e-compose-keep-self-cc nil
          mu4e-view-show-addresses t
          mu4e-hide-index-messages t
          mu4e-headers-results-limit 2000 ; this causes a performance impact
          mu4e-maildir-shortcuts
          '( ("/work/inbox"           . ?i)
             ("/work/archive"         . ?a)
             ("/work/drafts"          . ?d)
             ("/personal/INBOX"       . ?I)
             ("/personal/Archive"     . ?A)
             ("/personal/BulkInbox"   . ?b)
             ("/personal/BulkArchive" . ?B)
             ("/personal/drafts"      . ?D)))

    ;; contexts/multi-account
    (setq mu4e-contexts
          `( ,(make-mu4e-context
               :name "work"
               :enter-func (lambda () (mu4e-message "Switch to the work context"))
               ;; leave-fun not defined
               :match-func (lambda (msg)
                             (when msg
                               (mu4e-message-contact-field-matches msg
                                                                   :to "dsantana@domain.co.za")))
               :vars '(  ( user-mail-address         . "dsantana@domain.co.za" )
                         (mu4e-sent-folder . "/work/sent")
                         (mu4e-drafts-folder . "/work/drafts")
                         (mu4e-trash-folder . "/work/trash")
                         (mu4e-refile-folder . "/work/archive")
                         (smtpmail-default-smtp-server . "smtp.office365.com")
                         (smtpmail-smtp-user . "myusername@domain.co.za")
                         (smtpmail-smtp-server . "smtp.office365.com")
                         (smtpmail-local-domain . "domain.co.za")
                         (smtpmail-smtp-service . 587)
                         (message-signature-file . "~/work.signature")))
             ,(make-mu4e-context
               :name "personal"
               :enter-func (lambda () (mu4e-message "Switch to the personal context"))
               ;; leave-func not defined
               :match-func (lambda (msg)
                             (when msg
                               (mu4e-message-contact-field-matches msg
                                                                   :to "divan@home_domain.co.za")))
               :vars '(  ( user-mail-address         . "divan@home_domain.co.za"  )
                         (mu4e-reply-to-address . "divan@home_domain.co.za")
                         (mu4e-sent-folder . "/personal/Sent")
                         (mu4e-trash-folder . "/personal/Trash")
                         (mu4e-drafts-folder . "/personal/Drafts")
                         (mu4e-refile-folder . "/personal/Archive")
                         (smtpmail-default-smtp-server . "mail.home_domain.co.za")
                         (smtpmail-smtp-user . "divan@home_domain.co.za")
                         (smtpmail-smtp-server . "mail.home_domain.co.za")
                         (smtpmail-local-domain . "home_domain.co.za")
                         (smtpmail-smtp-service . 587)
                         ( message-signature-file . "~/personal.signature")))))

    (setq mu4e-context-policy 'pick-first)
    (setq mu4e-compose-context-policy nil)
    )
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(package-selected-packages (quote (use-package))))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )
```

and with this password entry:
```
$ pass show smtp.office365.com
mypassword
```

results in taking really long to send an email, as it has to do a deep search
as per below output. Sometimes the email gets sent, sometimes it doesn't.
```
Sending...
Mark set [2 times]
Sending via mail...
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-password-store: searching for ’smtp.office365.com’ in entry names not including user
auth-password-store: no entry name matched ’smtp.office365.com’, looking inside entries
auth-source-search: found 0 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-password-store: searching for ’smtp.office365.com’ in entry names not including user
auth-password-store: no entry name matched ’smtp.office365.com’, looking inside entries
auth-source-search: found 0 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-password-store: searching for ’smtp.office365.com’ in entry names not including user
auth-password-store: no entry name matched ’smtp.office365.com’, looking inside entries
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-source-search: found 1 results (max 1) matching (:max 1 :host "smtp.office365.com" :port "587")
auth-source-search: CREATED 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
Save auth info to file ~/.authinfo? [y/n/N/e/?] n
smtpmail-send-command: Process smtpmail not running
```

Trying the same as above with `massive-refactor` branch results in the message
failing to send. Output:
```
Sending via mail...
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
```

Doing the same as above, with `massive-refactor` branch and a slightly different entry, one I'd prefer, having just the base domain like below:
```
$ pass show office365.com
mypassword
```

results in it finding the entry, though with issues and failing to send.
```
Sending via mail...
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
```

Doing the same as above though, including the `user` field in the pass
entry, like so
```
$ pass show office365.com
mypassword
user:"myusername@domain.co.za"
```
also results in email failing to send with
```
Sending via mail...
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-command: Process smtpmail not running
```

Another test, same as above, this time with the pass entry containing the full host name:
```
$ pass show smtp.office365.com
mypassword
user:"myusername@domain.co.za"
```
same result:
```
Sending via mail...
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’myusername@domain.co.za@smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: myusername@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-command: Process smtpmail not running
```
Another test, this time with master branch, the full host as the pass entry, containeing the username too:

```
$ pass show smtp.office365.com
mypassword
user:"myusername@domain.co.za"
```

```
Sending via mail...
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-password-store: searching for ’smtp.office365.com’ in entry names not including user
auth-password-store: no entry name matched ’smtp.office365.com’, looking inside entries
auth-source-search: found 0 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-password-store: searching for ’smtp.office365.com’ in entry names not including user
auth-password-store: no entry name matched ’smtp.office365.com’, looking inside entries
auth-source-search: found 0 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-password-store: searching for ’smtp.office365.com’ in entry names not including user
auth-password-store: no entry name matched ’smtp.office365.com’, looking inside entries
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-source-search: found 1 results (max 1) matching (:max 1 :host "smtp.office365.com" :port "587")
auth-source-search: CREATED 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "myusername@domain.co.za" :max 1 :require (:user :secret) :create t)
Save auth info to file ~/.authinfo? [y/n/N/e/?] n
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
```
Also fails.

Ok last time, I think, same as above, but don't specify the smtpmail-smtp-user
in the `init.el` config, only in the pass auth entry as per previous example, also fails.
```
Sending via mail...
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user nil :max 1 :require nil :create nil)
auth-password-store: searching for ’smtp.office365.com’ in entry names including user
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user nil :max 1 :require (:user :secret) :create t)
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
Mark set [2 times]
```

Weird, because at some stage I had my email sending working with pass.

Sorry for the long winded report. Hopefully it helps testing.

**Comment by DamienCassou** [comment]
Thank you very much for your detailed report. A few things I noticed:

1. the username contains a domain (`someone@someplace.com`): this is something I didn't pay attention to and I will have to fix it
1. when nothing is found with `smtp.office365.com`, auth-password-store must start the search from scratch with `office365.com` (currently, only a subset of the search is done). I will have to fix that as well
1. in many cases, a password is found (`found 1 results`) but your email doesn't get sent. This is weird. I guess `auth-password-store` doesn't return what is expected by `smtpmail-send-it`.
1. can you please try with this entry (the same as yours, but without the double quotes surround user):
    ```
    $ pass show smtp.office365.com
    mypassword
    user:myusername@domain.co.za
    ```
1. Could you please set these 2 variables, this will help debugging:
    ```lisp
    (setq auth-source-do-cache nil)
    (setq auth-sources '(password-store))
    ```

I will do the changes in 1 and 2, but I'm still interested in what happens when you do 4 and 5 and do again some more experiments. Thanks for your help.

**Comment by DamienCassou** [comment]
I've updated `massive-refactor`. Please try again but do 4 and 5 before.

**Comment by divansantana** [comment]
> 4. can you please try with this entry (the same as yours, but without the double
> quotes surround user):
> ```
> $ pass show smtp.office365.com mypassword
> user:myusername@domain.co.za
> ```

Actually I made a mistake in my previous report. The user field already _wasn't_ quoted.

> Could you please set these 2 variables, this will help debugging: lisp (setq
> auth-source-do-cache nil) (setq auth-sources '(password-store))

Ok, I've set the above and tried again using `massive-refactor`(before your
latest commits), my pass entry looks like this:

```
 ᐅ pass show smtp.office365.com                    
mypassword
user:username@domain.co.za
 ᐅ 
```
results in email being failed to send with below:
```
Sending via mail...
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
```

I did, double check that those two new vars you gave me _are_ being set as you
requested.

Testing the same as above, this time with your latest commits in `massive-refactor` results in similar result:
```
Sending via mail...
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-
source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
```

**Comment by DamienCassou** [comment]
I've just push a new version to `massive-refactor`. It should print a bit more information.

**Comment by divansantana** [comment]
Thanks. I've been meaning to try look into how to get more verbosity out of
smtpmail and debug this more, just haven't got a chance yet.

Results with your latest commits in `massive-refactor`:

With this pass entry, same old emacs config.

```
 ᐅ pass show smtp.office365.com
mypassword
user:username@domain.co.za
```
Results in this output:
```
Sending...
Mark set [2 times]
Sending via mail...
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-password-store: return (:host smtp.office365.com :port 587 :user nil :secret (closure ((entry . smtp.office365.com) (port . 587) (user . username@domain.co.za) (host . smtp.office365.com) (type . password-store) (backend . [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search]) (spec :backend [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search] :type password-store :max 1 :require nil :create nil :delete nil :host smtp.office365.com :port 587 :user username@domain.co.za :max 1 :require nil :create nil) t) nil (auth-pass-get (quote secret) entry))) as final result (password is: nil)
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-password-store: return (:host smtp.office365.com :port 587 :user nil :secret (closure ((entry . smtp.office365.com) (port . 587) (user . username@domain.co.za) (host . smtp.office365.com) (type . password-store) (backend . [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search]) (spec :backend [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search] :type password-store :max 1 :require (:user :secret) :create nil :delete nil :host smtp.office365.com :port 587 :user username@domain.co.za :max 1 :require (:user :secret) :create t) t) nil (auth-pass-get (quote secret) entry))) as final result (password is: nil)
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-it: Sending failed: 530 5.7.57 SMTP; Client was not authenticated to send anonymous mail during MAIL FROM
```

**Comment by DamienCassou** [comment]
hmmm, that's strange. At the end of the last `auth-password-store` line, I can see `as final result (password is: nil)`. At the beginning, I can see `:user nil`. Both of these are equally strange. Can you please tell me the results of:

- `M-x (auth-pass-get "user" "smtp.office365.com")`
- `M-x (auth-pass-get 'secret "smtp.office365.com")`

**Comment by divansantana** [comment]
Thanks those commands helped debug the issue.

It seems one of the issues is a fault of my testing... big apologies. :(

Initially I think I was doing the reporting with my vanilla/debugging emacs
config launched from my normal home directory.

Then I got fancy to try ease debugging and had another emacs instance in
`~/dotfiles/emacstesting` of which I launched with `env
HOME=/home/admin/dotfiles/emacstesting emacs`. So I could continue to run and
document the findings with my standard "full" emacs in parallel.

I thought I _did_ symlink all that was necessary into `~/dotfiles/emacstesting`
to make all work, but apparently not:

Like so:
```
~/dotfiles/emacstesting  ᐅ ls -l1a
total 24
drwxr-xr-x  6 admin admin 4096 Oct 15 22:53 .
drwxr-xr-x 12 admin admin 4096 Oct 20 16:06 ..
lrwxrwxrwx  1 admin admin   19 Oct 15 22:53 .authinfo.gpg -> ../../.authinfo.gpg
drwxr-xr-x  2 admin admin 4096 Oct 15 17:44 .cache
drwx------  3 admin admin 4096 Oct 15 17:44 .config
drwxr-xr-x  5 admin admin 4096 Oct 17 21:59 .emacs.d
drwx------  2 admin admin 4096 Oct 15 17:44 .gnupg
lrwxrwxrwx  1 admin admin   10 Oct 15 17:46 mail -> ../../mail
lrwxrwxrwx  1 admin admin    9 Oct 15 17:46 .mu -> ../../.mu
lrwxrwxrwx  1 admin admin   18 Oct 15 18:06 .offlineimap -> ../../.offlineimap
lrwxrwxrwx  1 admin admin   20 Oct 15 18:06 .offlineimaprc -> ../../.offlineimaprc
lrwxrwxrwx  1 admin admin   21 Oct 15 17:43 .password-store -> ../../.password-store
```

Pass did find the entry, as below shows
```
~/dotfiles/emacstesting  ᐅ env HOME=/home/admin/dotfiles/emacstesting pass |grep smtp.office365.com
├── smtp.office365.com
```

The issue is something to do with my gpg decryption not working from that
environment, hence the return of `nil` from `(auth-pass-get 'secret
"smtp.office365.com")` and `(auth-pass-get "user" "smtp.office365.com")`:
```
~/dotfiles/emacstesting  ᐅ env HOME=/home/admin/dotfiles/emacstesting pass show smtp.office365.com
gpg: decryption failed: No secret key
~/dotfiles/emacstesting  ᐅ
```

Getting to fancy with the bug reporting, bit me. Sorry about that and the time
wasted.

Ok so back to the issue. Now testing it properly with the vanilla config
installed in my home.

Result, all works, quick email sent, with below output.
```
Sending via mail...
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-password-store: return (:host smtp.office365.com :port 587 :user username@domain.co.za :secret (closure ((entry . smtp.office365.com) (port . 587) (user . username@domain.co.za) (host . smtp.office365.com) (type . password-store) (backend . [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search]) (spec :backend [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search] :type password-store :max 1 :require nil :create nil :delete nil :host smtp.office365.com :port 587 :user username@domain.co.za :max 1 :require nil :create nil) t) nil (auth-pass-get (quote secret) entry))) as final result (password is: mypassword)
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require nil :create nil)
Sending email
Sending email done
Mark set
Saving file /home/admin/mail/work/sent/cur/1477039402.1ccc0924117e4acd.laptop:2,S...
Wrote /home/admin/mail/work/sent/cur/1477039402.1ccc0924117e4acd.laptop:2,S
Sending...done
[mu4e] Message sent
```

The above is with a pass entry of:
```
~ ᐅ pass show smtp.office365.com
mypassword
user:username@domain.co.za
```

Ok now trying same as above, without the `user` field bit in the pass entry like so
```
~ ᐅ pass show smtp.office365.com
mypassword
```

Results in the familiar failed email message:
```
Sending...
Mark set [2 times]
Sending via mail...
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-password-store: return (:host smtp.office365.com :port 587 :user nil :secret (closure ((entry . smtp.office365.com) (port . 587) (user . username@domain.co.za) (host . smtp.office365.com) (type . password-store) (backend . [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search]) (spec :backend [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search] :type password-store :max 1 :require nil :create nil :delete nil :host smtp.office365.com :port 587 :user username@domain.co.za :max 1 :require nil :create nil) t) nil (auth-pass-get (quote secret) entry))) as final result (password is: mypassword)
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require nil :create nil)
auth-password-store: searching for ’username@domain.co.za@smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: no match found
auth-password-store: searching for ’smtp.office365.com’ in entry names (user: username@domain.co.za)
auth-password-store: found 1 match: smtp.office365.com
auth-password-store: return (:host smtp.office365.com :port 587 :user nil :secret (closure ((entry . smtp.office365.com) (port . 587) (user . username@domain.co.za) (host . smtp.office365.com) (type . password-store) (backend . [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search]) (spec :backend [eieio-class-tag--auth-source-backend password-store . t t t nil ignore auth-pass-search] :type password-store :max 1 :require (:user :secret) :create nil :delete nil :host smtp.office365.com :port 587 :user username@domain.co.za :max 1 :require (:user :secret) :create t) t) nil (auth-pass-get (quote secret) entry))) as final result (password is: mypassword)
auth-source-search: found 1 results (max 1) matching (:host "smtp.office365.com" :port "587" :user "username@domain.co.za" :max 1 :require (:user :secret) :create t)
smtpmail-send-command: Process smtpmail not running
```

Suppose that's because `(auth-pass-get "user" "smtp.office365.com")` correctly
returns `nil` since the user field is absent in the entry.

Though Emacs already knows the username, and therefore if the user field is not
defined in pass(returns nil), perhaps it should return the original username and
not nil, so the email will still send? Think that's the issue with this one.

**Comment by DamienCassou** [comment]
Don't worry, we all make mistakes. I've pushed a new version. Could you please try again after removing the "user" field from the entry?

**Comment by DamienCassou** [comment]
Thank you very much for your help. I've just released v2.0.0.

**Comment by divansantana** [comment]
@DamienCassou great. Thanks for a good release.
